text/0018-open-context-provider-framework.md

```diff
@@ -0,0 +1,157 @@
+synthesis. This increases the size of the Cfn template generated by the app, increases the time required to deploy the stack, adds unnecessary
+extra resources to the user's account, and makes it possible for a value generated by the CfnCustomResource to differ from one deployment
+to the next.
+
+Unfortunately, these CfnCustomResources required by 3rd party construct libraries may also be a security risk to the user. The lambda that
+backs them persists as long as the stack is deployed, and that lambda is granted permissions into the account that may be possible to leverage
+by an attacker that gains access to the lambda; depending on the lambda's purpose, of course.
+
+# Basic Example
+
+```ts
+// Allow the provider to be an arbitrary string
+const someValues: thirdpartlib.DoSomethingContextReponse = ContextProvider.getValue(scope, {
+  provider: 'thidparty-do-something',
+  props: {
+    // whatever needed
+  },
+  dummyValue: undefined,
+}).value;
+// do something with someValues -- which has been materialized at synth time.
+
+// Still retain the existing functionality (ex: Vpc.fromLookup())
+const attributes: cxapi.VpcContextResponse = ContextProvider.getValue(scope, {
+  provider: cxschema.ContextProvider.VPC_PROVIDER,
+  props: {
+    filter,
+    returnAsymmetricSubnets: true,
+    subnetGroupNameTag: options.subnetGroupNameTag,
+  } as cxschema.VpcContextQuery,
+  dummyValue: undefined,
+}).value;
+```
+
+# Design Summary
+
+Leverage the existing `Plugin` mechanism that is implemeted into the CDK CLI, and currently only used to register `CredentialProvider`s.
+Extend the `PluginHost` class to provide a method for registering a `ContextProvider` plugin.
+
+# Detailed Design
+
+## Background -- Understanding the current CDK structure for ContextProvider
+
+CDK CLI: `packages/aws-cdk`
+
+CDK CLI entrypoint: `bin/cdk.ts`: `initCommandLine()`
+
+1. Load a configuration (`Configuration().load()` -- `lib/settings.ts`) -- merges key:value pairs from cdk.json into settings
+2. Load the plugins from the settings['plugin'] key (i.e. 'plugin' key from cdk.json)
+
+Loading plugins -- happens in `initCommandLine()`
+
+* Resolves the plugin module, and then invokes `PluginHost.instance.load(plugin)` ( `lib/plugin.ts` )
+* `lib/plugin.ts` defines `interface Plugin { version: '1'; init?: (host: PluginHost) => void; }`
+* `PluginHost.load(moduleSpec: string)` -- if arg is a plugin, then run `plugin.init(PluginHost.instance)`.
+* `plugin.init(PluginHost.instance)` can invoke `PluginHost.instance.registerCredentialProviderSource` to register itself
+* Currently, only registration method is `registerCredentialProviderSource`
+
+ContextProvider -- `context-providers/`
+
+* All of the context provider implementations are baked into the CDK CLI app.
+* Base: `interface ContextProviderPlugin { getValue(args: {[key: string]: any}): Promise<any>; }` (in `context-providers/provider.ts`)
```

Comment:
Let's talk AWS permissions.

Don't forget that the specific instances that implement `ContextProviderPlugin` have also been initialized with `SdkProvider`s, which allow them to get AWS credentials for any account/region pair.

The context provider is going to have to do AWS calls. Where is it going to get access to an SDK instance? How are we communicating that? How do we define the API between plugin and CDK? Who initializes it? I'm not a fan of just passing an `SdkProvider`, both for "security" reasons as well as keeping the API surface as minimal as we can keep it and not leaking implementation details.

There needs to be a TypeScript package somewhere that contains *just* the API definitions for this plugin relationship. What are the types that are going to go in there? Just "whatever the AWS SDK for JS exports"? Or are we going to wrap those clients?

To sidestep that--maybe we just say that external context providers are subprocesses, and we initialize the environment with `AWS_REGION` and `AWS_ACCESS_KEY_ID` (etc)?

------

As an additional wrinkle: right now, we just run the context providers using the "current" AWS credentials, and it's the user's responsibility to make sure those are sufficient for whatever it is they intend to do.  In the near future, we will introduce a special role that will be bootstrapped into the target account you will be querying for context in. That role will have a fixed set of permissions, and so what role should your `CUSTOM` context provider assume? If it wants to do, say, `s3:ListBuckets`, that role we provide by default won't have the permissions to do that.

At the same time, I'm not sure we can justify attaching `aws/ReadOnlyAccess` to that role, so I guess the user will just manually have to add policy statements at bootstrap time to satisfy the needs of the context providers they intend to be using?

Reply:
The context provider could specify the role that it needs (e.g., a `PolicyStatement`), which the CDK could bootstrap (and validate, if we want to ensure it doesn't require any write permissions). Further, the CDK could ask the user to authorize the creation of those roles (unless `--require-approval never` is specified).

The context provider would be created with an SDK from `SdkProvider::withAssumedRole`. If there's a reason not to expose the full SDK, there could be a package exposing all or some services. It's not clear to me there's any value in such a thin wrapper of the SDK.